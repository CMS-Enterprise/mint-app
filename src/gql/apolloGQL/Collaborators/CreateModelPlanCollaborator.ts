import { gql } from '@apollo/client';

export default gql(/* GraphQL */ `
  mutation CreateModelPlanCollaborator($input: PlanCollaboratorCreateInput!) {
    createPlanCollaborator(input: $input) {
      teamRoles
      userAccount {
        id
        commonName
        email
      }
      userID
      modelPlanID
    }
  }
`);
