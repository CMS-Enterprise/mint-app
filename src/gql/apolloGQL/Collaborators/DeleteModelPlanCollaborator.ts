import { gql } from '@apollo/client';

export default gql(/* GraphQL */ `
  mutation DeleteModelPlanCollaborator($id: UUID!) {
    deletePlanCollaborator(id: $id) {
      id
      teamRoles
      userAccount {
        id
        commonName
        email
        username
      }
      userID
      modelPlanID
    }
  }
`);
