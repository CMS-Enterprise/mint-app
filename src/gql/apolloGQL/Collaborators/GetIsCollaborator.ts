import { gql } from '@apollo/client';

export default gql(/* GraphQL */ `
  query GetIsCollaborator($id: UUID!) {
    modelPlan(id: $id) {
      id
      isCollaborator
    }
  }
`);
