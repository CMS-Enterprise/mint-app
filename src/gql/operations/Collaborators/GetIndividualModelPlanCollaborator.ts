import { gql } from '@apollo/client';

export default gql(/* GraphQL */ `
  query GetIndividualModelPlanCollaborator($id: UUID!) {
    planCollaboratorByID(id: $id) {
      id
      userAccount {
        id
        commonName
        email
        username
      }
      userID
      teamRoles
    }
  }
`);
