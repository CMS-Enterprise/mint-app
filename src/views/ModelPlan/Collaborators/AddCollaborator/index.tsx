import React, { useRef } from 'react';
import { Trans, useTranslation } from 'react-i18next';
import { useHistory, useLocation, useParams } from 'react-router-dom';
import { Button, Fieldset, Label, TextInput } from '@trussworks/react-uswds';
import classNames from 'classnames';
import { Field, Form, Formik, FormikProps } from 'formik';
import {
  GetIndividualModelPlanCollaboratorQuery,
  GetModelCollaboratorsQuery,
  TeamRole,
  useCreateModelPlanCollaboratorMutation,
  useGetModelCollaboratorsQuery,
  useUpdateModelPlanCollaboratorMutation
} from 'gql/gen/graphql';

import Breadcrumbs, { BreadcrumbItemOptions } from 'components/Breadcrumbs';
import UswdsReactLink from 'components/LinkWrapper';
import MainContent from 'components/MainContent';
import OktaUserSelect from 'components/OktaUserSelect';
import PageHeading from 'components/PageHeading';
import Alert from 'components/shared/Alert';
import { ErrorAlert, ErrorAlertMessage } from 'components/shared/ErrorAlert';
import FieldErrorMsg from 'components/shared/FieldErrorMsg';
import FieldGroup from 'components/shared/FieldGroup';
import MultiSelect from 'components/shared/MultiSelect';
import RequiredAsterisk from 'components/shared/RequiredAsterisk';
import Spinner from 'components/Spinner';
import useMessage from 'hooks/useMessage';
import usePlanTranslation from 'hooks/usePlanTranslation';
import { getKeys } from 'types/translation';
import flattenErrors from 'utils/flattenErrors';
import { composeMultiSelectOptions } from 'utils/modelPlan';
import CollaboratorsValidationSchema from 'validations/modelPlanCollaborators';

import RoleInfo from '../_components/RoleInfo';
import { isLastModelLead } from '..';

type GetCollaboratorsType =
  GetModelCollaboratorsQuery['modelPlan']['collaborators'][0];
type CollaboratorFormType =
  GetIndividualModelPlanCollaboratorQuery['planCollaboratorByID'];

type LocationProps = {
  fromCollaborationArea: boolean;
  pathname: string;
  state: {
    fromCollaborationArea: boolean;
  };
};

const Collaborators = () => {
  const { t: collaboratorsT } = useTranslation('collaborators');
  const { t: collaboratorsMiscT } = useTranslation('collaboratorsMisc');
  const { t: miscellaneousT } = useTranslation('miscellaneous');
  const { teamRoles: teamRolesConfig } = usePlanTranslation('collaborators');

  const history = useHistory();

  const location = useLocation<LocationProps>();

  const isFromCollaborationArea = location.state?.fromCollaborationArea;

  const params = new URLSearchParams(location.search);

  const manageOrAdd = params.get('view') || 'manage';

  const { showMessageOnNextPage } = useMessage();

  const { modelID, collaboratorId } = useParams<{
    modelID: string;
    collaboratorId: string;
  }>();

  const formikRef = useRef<FormikProps<CollaboratorFormType>>(null);

  const [create, { loading }] = useCreateModelPlanCollaboratorMutation();

  const [update, { loading: updateLoading }] =
    useUpdateModelPlanCollaboratorMutation();

  const { data: allCollaboratorsData, loading: queryLoading } =
    useGetModelCollaboratorsQuery({
      variables: {
        id: modelID
      }
    });

  const allCollaborators =
    allCollaboratorsData?.modelPlan?.collaborators ??
    ([] as GetCollaboratorsType[]);

  const specificCollaborator = allCollaborators.filter(
    collab => collab.id === collaboratorId
  );

  const initialValues: CollaboratorFormType = specificCollaborator[0] ?? {
    userAccount: {}
  };

  const isModelLead = specificCollaborator[0]?.teamRoles?.includes(
    TeamRole.MODEL_LEAD
  );

  const handleUpdateDraftModelPlan = (formikValues?: CollaboratorFormType) => {
    const {
      userAccount: { username, commonName },
      teamRoles
    } = formikValues || { userAccount: { userName: null } };
    if (collaboratorId) {
      update({
        variables: {
          id: collaboratorId,
          newRole: teamRoles!
        }
      })
        .then(response => {
          if (!response?.errors) {
            showMessageOnNextPage(
              <>
                <Alert
                  type="success"
                  slim
                  data-testid="success-collaborator-alert"
                  className="margin-y-4"
                >
                  {collaboratorsMiscT('successUpdateMessage', {
                    collaborator: commonName,
                    role: teamRoles
                      ?.map((role: TeamRole) => {
                        return collaboratorsT(`teamRoles.options.${role}`);
                      })
                      .join(', ')
                  })}
                </Alert>
              </>
            );
            history.push(
              `/models/${modelID}/collaboration-area/collaborators?view=${manageOrAdd}`
            );
          }
        })
        .catch(errors => {
          formikRef?.current?.setErrors(errors);
        });
    } else {
      create({
        variables: {
          input: {
            modelPlanID: modelID,
            userName: username!,
            teamRoles: teamRoles!
          }
        }
      })
        .then(response => {
          if (!response?.errors) {
            showMessageOnNextPage(
              <>
                <Alert
                  type="success"
                  slim
                  data-testid="success-collaborator-alert"
                  className="margin-y-4"
                >
                  {collaboratorsMiscT('successMessage', {
                    collaborator: commonName,
                    role: teamRoles
                      ?.map((role: TeamRole) => {
                        return collaboratorsT(`teamRoles.options.${role}`);
                      })
                      .join(', ')
                  })}
                </Alert>
              </>
            );
            history.push(
              `/models/${modelID}/collaboration-area/collaborators?view=${manageOrAdd}`
            );
          }
        })
        .catch(errors => {
          const collaboratorExistingError =
            errors.graphQLErrors[0]?.message.includes(
              'unique_collaborator_per_plan'
            );
          if (collaboratorExistingError) {
            formikRef?.current?.setErrors({
              userAccount: {
                username: collaboratorsMiscT('existingMember')
              }
            });
          } else {
            formikRef?.current?.setErrors(errors);
          }
        });
    }
  };

  const breadcrumbs = [BreadcrumbItemOptions.HOME];

  if (manageOrAdd === 'manage') {
    breadcrumbs.push(BreadcrumbItemOptions.COLLABORATION_AREA);
  }

  if (!isFromCollaborationArea || manageOrAdd === 'add') {
    breadcrumbs.push(BreadcrumbItemOptions.COLLABORATORS);
  }

  return (
    <MainContent>
      <div className="grid-container">
        <Breadcrumbs
          items={breadcrumbs}
          customItem={
            collaboratorId
              ? collaboratorsMiscT('collaboratorsMisc:updateATeamMember')
              : collaboratorsMiscT('collaboratorsMisc:addATeamMember')
          }
        />
        <div className="desktop:grid-col-6">
          <PageHeading className="margin-top-6 margin-bottom-2">
            {collaboratorId
              ? collaboratorsMiscT('updateATeamMember')
              : collaboratorsMiscT('addATeamMember')}
          </PageHeading>

          {!collaboratorId && (
            <div className="margin-bottom-2 font-body-md line-height-body-5">
              {!collaboratorId && collaboratorsMiscT('searchTeamInfo')}{' '}
              {collaboratorsMiscT('teamInfo')}
            </div>
          )}

          <p className="margin-bottom-5">
            <Trans
              i18nKey={miscellaneousT('allFieldsRequired')}
              components={{
                s: <span className="text-secondary-dark" />
              }}
            />
          </p>

          <Formik
            initialValues={initialValues}
            enableReinitialize
            onSubmit={handleUpdateDraftModelPlan}
            validationSchema={CollaboratorsValidationSchema}
            validateOnBlur={false}
            validateOnChange={false}
            validateOnMount={false}
            innerRef={formikRef}
          >
            {(formikProps: FormikProps<CollaboratorFormType>) => {
              const { errors, values, setFieldValue, handleSubmit } =
                formikProps;
              const flatErrors = flattenErrors(errors);

              return (
                <>
                  {getKeys(errors).length > 0 && (
                    <ErrorAlert
                      testId="formik-validation-errors"
                      classNames="margin-top-3"
                      heading={miscellaneousT('checkAndFix')}
                    >
                      {getKeys(flatErrors).map(key => {
                        return (
                          <ErrorAlertMessage
                            key={`Error.${key}`}
                            errorKey={`${key}`}
                            message={flatErrors[key]}
                          />
                        );
                      })}
                    </ErrorAlert>
                  )}

                  <Form
                    onSubmit={e => {
                      handleSubmit(e);
                      window.scrollTo(0, 0);
                    }}
                  >
                    <Fieldset disabled={queryLoading}>
                      <FieldGroup
                        className="margin-top-0"
                        scrollElement="userAccount.commonName"
                        error={!!flatErrors['userAccount.commonName']}
                      >
                        <Label
                          htmlFor="model-team-cedar-contact"
                          id="label-model-team-cedar-contact"
                        >
                          {collaboratorsT('username.label')}
                          <RequiredAsterisk />
                        </Label>

                        <FieldErrorMsg>
                          {flatErrors['userAccount.commonName']}
                        </FieldErrorMsg>

                        <Label
                          id="hint-model-team-cedar-contact"
                          htmlFor="model-team-cedar-contact"
                          className="text-normal margin-top-1 margin-bottom-105 text-base maxw-none"
                          hint
                        >
                          {collaboratorsMiscT('startTyping')}
                        </Label>

                        {collaboratorId ? (
                          <Field
                            as={TextInput}
                            disabled
                            error={!!flatErrors['userAccount.commonName']}
                            className="margin-top-1"
                            id="collaboration-full-name"
                            name="userAccount.commonName"
                          />
                        ) : (
                          <OktaUserSelect
                            id="model-team-cedar-contact"
                            name="model-team-cedar-contact"
                            ariaLabelledBy="label-model-team-cedar-contact"
                            ariaDescribedBy="hint-model-team-cedar-contact"
                            onChange={oktaUser => {
                              setFieldValue(
                                'userAccount.commonName',
                                oktaUser?.displayName
                              );
                              setFieldValue(
                                'userAccount.username',
                                oktaUser?.username
                              );
                            }}
                          />
                        )}
                      </FieldGroup>

                      <FieldGroup
                        scrollElement="teamRoles"
                        error={!!flatErrors.teamRoles}
                      >
                        <Label htmlFor="collaborator-role">
                          {collaboratorsT('teamRoles.label')}
                          <RequiredAsterisk />
                        </Label>

                        <FieldErrorMsg>{flatErrors.teamRoles}</FieldErrorMsg>

                        <Field
                          as={MultiSelect}
                          id="collaborator-role"
                          name="role"
                          selectedLabel={collaboratorsMiscT('selectedRoles')}
                          options={composeMultiSelectOptions(
                            teamRolesConfig.options,
                            undefined,
                            isModelLead && isLastModelLead(allCollaborators)
                              ? TeamRole.MODEL_LEAD
                              : ''
                          )}
                          onChange={(value: TeamRole[]) => {
                            setFieldValue('teamRoles', value);
                          }}
                          initialValues={initialValues.teamRoles}
                          tagOrder={
                            teamRolesConfig.options[TeamRole.MODEL_LEAD]
                          }
                          disabledOption={
                            isModelLead && isLastModelLead(allCollaborators)
                          }
                          disabledLabel={
                            teamRolesConfig.options[TeamRole.MODEL_LEAD]
                          }
                        />
                      </FieldGroup>

                      <RoleInfo
                        className={classNames('margin-bottom-5', {
                          'margin-bottom-6': collaboratorId
                        })}
                      />

                      {!collaboratorId && (
                        <Alert
                          type="info"
                          slim
                          data-testid="mandatory-fields-alert"
                          className="margin-y-3"
                        >
                          <span className="mandatory-fields-alert__text">
                            {isModelLead && isLastModelLead(allCollaborators)
                              ? collaboratorsMiscT('lastModelLeadMemberInfo')
                              : collaboratorsMiscT('searchMemberInfo')}
                          </span>
                        </Alert>
                      )}

                      <div className="margin-y-3 display-block">
                        <Button
                          type="submit"
                          disabled={
                            !values.userAccount.commonName || !values.teamRoles
                          }
                        >
                          {!collaboratorId
                            ? collaboratorsMiscT('addTeamMemberButton')
                            : collaboratorsMiscT('saveChanges')}
                        </Button>

                        {(loading || updateLoading) && (
                          <Spinner className="margin-left-2" />
                        )}
                      </div>
                    </Fieldset>
                  </Form>
                </>
              );
            }}
          </Formik>

          <UswdsReactLink
            to={
              isFromCollaborationArea
                ? `/models/${modelID}/collaboration-area`
                : `/models/${modelID}/collaboration-area/collaborators?view=${manageOrAdd}`
            }
          >
            <span>&larr; </span>{' '}
            {!collaboratorId
              ? collaboratorsMiscT('dontAddTeamMember')
              : collaboratorsMiscT('dontUpdateTeamMember')}
          </UswdsReactLink>
        </div>
      </div>
    </MainContent>
  );
};

export default Collaborators;
