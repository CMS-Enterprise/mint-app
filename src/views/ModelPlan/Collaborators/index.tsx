import React, { useEffect, useMemo, useState } from 'react';
import { useTranslation } from 'react-i18next';
import { RootStateOrAny, useSelector } from 'react-redux';
import { Link, Route, Switch, useHistory, useParams } from 'react-router-dom';
import { useMutation, useQuery } from '@apollo/client';
import {
  Breadcrumb,
  BreadcrumbBar,
  BreadcrumbLink,
  Button,
  Grid,
  GridContainer
} from '@trussworks/react-uswds';

import UswdsReactLink from 'components/LinkWrapper';
import MainContent from 'components/MainContent';
import Modal from 'components/Modal';
import PageHeading from 'components/PageHeading';
import PageLoading from 'components/PageLoading';
import Alert from 'components/shared/Alert';
import Expire from 'components/shared/Expire';
import useMessage from 'hooks/useMessage';
import DeleteModelPlanCollaborator from 'queries/Collaborators/DeleteModelPlanCollaborator';
import GetModelPlanCollaborators from 'queries/Collaborators/GetModelCollaborators';
import {
  DeleteModelPlanCollaborator as DeleteModelPlanCollaboratorType,
  DeleteModelPlanCollaborator_deletePlanCollaborator as ModelPlanCollaboratorType
} from 'queries/Collaborators/types/DeleteModelPlanCollaborator';
import {
  GetModelCollaborators,
  GetModelCollaborators_modelPlan_collaborators as GetCollaboratorsType
} from 'queries/Collaborators/types/GetModelCollaborators';
import NotFound from 'views/NotFound';

import AddCollaborator from './AddCollaborator';
import CollaboratorsTable from './table';

// Checking if there is only one collaborator with role of MODEL_LEAD - can't edit or remove if so
const isLastModelLead = (collaborators: GetCollaboratorsType[]) => {
  const modelLeads = collaborators.filter(
    collaborator => collaborator.teamRole === 'MODEL_LEAD'
  );
  return !(modelLeads.length > 1);
};

const SuccessRemovalMessage = ({
  modelName
}: {
  modelName: string | undefined;
}) => {
  const { t } = useTranslation('newModel');
  return (
    <Alert
      type="success"
      data-testid="mandatory-fields-alert"
      className="margin-y-4"
      heading={t('success.heading', {
        modelName
      })}
    >
      <span className="mandatory-fields-alert__text">
        {t('success.message')}
      </span>
    </Alert>
  );
};

export const CollaboratorsContent = () => {
  const { modelID } = useParams<{ modelID: string }>();
  const { t: h } = useTranslation('draftModelPlan');
  const { t } = useTranslation('newModel');

  const history = useHistory();

  const { message, showMessageOnNextPage } = useMessage();

  const [isModalOpen, setModalOpen] = useState(false);
  const [isLastLead, setIsLastLead] = useState(false);

  const [
    removeCollaborator,
    setRemoveCollaborator
  ] = useState<ModelPlanCollaboratorType>();

  // Current user's EUA id - to warn about removing yourself from model plan
  const { euaId } = useSelector((state: RootStateOrAny) => state.auth);

  const [mutate] = useMutation<DeleteModelPlanCollaboratorType>(
    DeleteModelPlanCollaborator
  );

  const { error, data, refetch, loading } = useQuery<GetModelCollaborators>(
    GetModelPlanCollaborators,
    {
      variables: {
        id: modelID
      }
    }
  );

  const collaborators = useMemo(() => {
    return (data?.modelPlan?.collaborators ?? []) as GetCollaboratorsType[];
  }, [data?.modelPlan?.collaborators]);

  // Setting state of isLastLead to toggle edit/remove links
  useEffect(() => {
    setIsLastLead(isLastModelLead(collaborators));
  }, [collaborators]);

  const handleRemoveCollaborator = (
    collaborator: ModelPlanCollaboratorType
  ) => {
    mutate({
      variables: {
        id: collaborator.id
      }
    })
      .then(response => {
        if (!response?.errors) {
          setModalOpen(false);
          if (collaborator.userAccount.username === euaId) {
            showMessageOnNextPage(
              <SuccessRemovalMessage modelName={data?.modelPlan?.modelName} />
            );
            history.push('/');
          } else {
            refetch();
          }
        }
      })
      .catch(errors => {
        setModalOpen(false);
        refetch();
      });
  };

  // Modal control for removing a collaborator
  // Conditional rendering if attempting to remove yourself from model plan
  const RemoveCollaborator = () => {
    // i18n key for conditionally rendering text
    const selfOrCollaborator: string =
      removeCollaborator?.userAccount.username === euaId
        ? 'selfModal'
        : 'modal';

    return (
      <Modal
        isOpen={isModalOpen}
        closeModal={() => setModalOpen(false)}
        className="confirmation-modal"
      >
        <PageHeading
          headingLevel="h3"
          className="margin-top-neg-2 margin-bottom-1"
        >
          {t(`${selfOrCollaborator}.heading`, {
            collaborator: removeCollaborator?.userAccount.commonName
          })}
        </PageHeading>
        <p>{t(`${selfOrCollaborator}.subheading`)}</p>
        <Button
          type="button"
          className="margin-right-4"
          onClick={() =>
            removeCollaborator && handleRemoveCollaborator(removeCollaborator)
          }
        >
          {t(`${selfOrCollaborator}.confirm`)}
        </Button>
        <Button type="button" unstyled onClick={() => setModalOpen(false)}>
          {t(`${selfOrCollaborator}.no`)}
        </Button>
      </Modal>
    );
  };

  if (error) {
    return <div>{JSON.stringify(error)}</div>;
  }

  return (
    <MainContent>
      {RemoveCollaborator()}

      <GridContainer>
        <Grid row gap>
          <Grid desktop={{ col: 12 }}>
            {message && <Expire delay={45000}>{message}</Expire>}

            <BreadcrumbBar variant="wrap">
              <Breadcrumb>
                <BreadcrumbLink asCustom={Link} to="/">
                  <span>{h('home')}</span>
                </BreadcrumbLink>
              </Breadcrumb>
              <Breadcrumb current>{t('teamBreadcrumb')}</Breadcrumb>
            </BreadcrumbBar>
            <PageHeading className="margin-top-4 margin-bottom-2">
              {t('headingTeamMembers')}
            </PageHeading>
            <div className="font-body-lg margin-bottom-6">
              {t('teamMemberInfo')}
            </div>
            <h4 className="margin-bottom-1">{t('teamMembers')}</h4>
            <UswdsReactLink
              className="usa-button margin-bottom-2"
              variant="unstyled"
              to={`/models/${modelID}/collaborators/add-collaborator`}
            >
              {t('addTeamMemberButton')}
            </UswdsReactLink>

            {loading && <PageLoading />}
            {!loading &&
              (collaborators.length === 0 ? (
                <Alert type="info" heading={t('table.noCollaborators')} />
              ) : (
                <CollaboratorsTable
                  collaborators={collaborators}
                  setModalOpen={setModalOpen}
                  setRemoveCollaborator={setRemoveCollaborator}
                  isLastLead={isLastLead}
                />
              ))}

            <div className="margin-top-5 display-block">
              <UswdsReactLink
                data-testid="continue-to-tasklist"
                className="usa-button usa-button--outline"
                variant="unstyled"
                to={`/models/${modelID}/task-list`}
              >
                {collaborators.length > 0
                  ? h('continueToTaskList')
                  : t('continueWithoutAdding')}
              </UswdsReactLink>
            </div>
          </Grid>
        </Grid>
      </GridContainer>
    </MainContent>
  );
};

const Collaborators = () => {
  return (
    <Switch>
      <Route
        path="/models/:modelID/collaborators"
        exact
        render={() => <CollaboratorsContent />}
      />
      <Route
        path="/models/:modelID/collaborators/add-collaborator/:collaboratorId?"
        exact
        render={() => <AddCollaborator />}
      />

      {/* 404 */}
      <Route path="*" render={() => <NotFound />} />
    </Switch>
  );
};

export default Collaborators;
