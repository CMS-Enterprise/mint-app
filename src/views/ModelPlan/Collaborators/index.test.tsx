import React from 'react';
import { MemoryRouter, Route } from 'react-router-dom';
import { MockedProvider } from '@apollo/client/testing';
import { render, screen, waitFor } from '@testing-library/react';

import GetModelPlanCollaborators from 'queries/Collaborators/GetModelCollaborators';

import Collaborators from './index';

describe('Collaborator/Team Member page w/table', () => {
  const mocks = [
    {
      request: {
        query: GetModelPlanCollaborators,
        variables: { id: 'f11eb129-2c80-4080-9440-439cbe1a286f' }
      },
      result: {
        data: {
          modelPlan: {
            collaborators: [
              {
                id: '61c7b30c-969d-4dd4-b13b-a5065f43be43',
                modelPlanID: 'f11eb129-2c80-4080-9440-439cbe1a286f',
                euaUserID: 'ABCD',
                email: 'jdoe@gmail.com',
                fullName: 'John Doe',
                teamRole: 'MODEL_LEAD',
                createdDts: '2022-10-22T00:00:00Z'
              }
            ]
          }
        }
      }
    }
  ];
  it('displays a table with collaborator', async () => {
    render(
      <MemoryRouter
        initialEntries={[
          'models/new-plan/f11eb129-2c80-4080-9440-439cbe1a286f/collaborators'
        ]}
      >
        <MockedProvider mocks={mocks} addTypename={false}>
          <Route path="models/new-plan/:modelID/collaborators">
            <Collaborators />
          </Route>
        </MockedProvider>
      </MemoryRouter>
    );

    expect(await screen.findByRole('table')).toBeInTheDocument();

    await waitFor(() => {
      expect(screen.getByText('Model Lead')).toBeInTheDocument();
    });

    await waitFor(() => {
      expect(screen.getByText('John Doe')).toBeInTheDocument();
    });
  });

  it('matches snapshot', async () => {
    const { asFragment } = render(
      <MemoryRouter
        initialEntries={[
          'models/new-plan/f11eb129-2c80-4080-9440-439cbe1a286f/collaborators'
        ]}
      >
        <MockedProvider mocks={mocks} addTypename={false}>
          <Route path="models/new-plan/:modelID/collaborators">
            <Collaborators />
          </Route>
        </MockedProvider>
      </MemoryRouter>
    );
    await waitFor(() => {
      expect(asFragment()).toMatchSnapshot();
    });
  });
});
