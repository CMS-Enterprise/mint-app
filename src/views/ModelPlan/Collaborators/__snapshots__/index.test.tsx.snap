// Vitest Snapshot v1, https://vitest.dev/guide/snapshot.html

exports[`Collaborator/Team Member page w/table > matches snapshot 1`] = `
<DocumentFragment>
  <main
    class="mint-main-content"
    id="main-content"
    role="main"
    tabindex="-1"
  >
    <div
      class="grid-container"
      data-testid="gridContainer"
    >
      <div
        class="grid-row grid-gap"
        data-testid="grid"
      >
        <div
          class="desktop:grid-col-12"
          data-testid="grid"
        >
          <nav
            aria-label="Breadcrumbs"
            class="usa-breadcrumb usa-breadcrumb--wrap"
          >
            <ol
              class="usa-breadcrumb__list"
            >
              <li
                class="usa-breadcrumb__list-item"
              >
                <a
                  class="usa-link usa-breadcrumb__link"
                  href="/"
                >
                  <span>
                    Home
                  </span>
                </a>
              </li>
              <li
                aria-current="page"
                class="usa-breadcrumb__list-item usa-current"
              >
                <span>
                  Add model team members
                </span>
              </li>
            </ol>
          </nav>
          <h1
            aria-live="polite"
            class="mint-h1 margin-top-4 margin-bottom-2"
            tabindex="-1"
          >
            Add model team members
          </h1>
          <div
            class="font-body-lg margin-bottom-6"
          >
            Add any team members who will be collaborating on your Model Plan. You can edit this list at any time.
          </div>
          <h4
            class="margin-bottom-1"
          >
            Team members
          </h4>
          <a
            class="usa-button margin-bottom-2"
            href="/models/f11eb129-2c80-4080-9440-439cbe1a286f/collaborators/add-collaborator?view=add"
          >
            Add team member
          </a>
          <div
            class="collaborator-table"
          >
            <table
              class="usa-table usa-table--borderless width-full"
              data-testid="table"
            >
              <caption
                class="usa-sr-only"
              >
                requestsTable.caption
              </caption>
              <thead>
                <tr
                  role="row"
                >
                  <th
                    aria-sort="none"
                    class="table-header"
                    colspan="1"
                    role="columnheader"
                    scope="col"
                    style="padding-bottom: .5rem;"
                  >
                    <button
                      class="usa-button usa-button--unstyled position-relative"
                      style="cursor: pointer;"
                      title="Toggle SortBy"
                      type="button"
                    >
                      Name
                      <svg
                        class="usa-icon margin-left-05 position-absolute"
                        data-testid="caret--sort"
                        focusable="false"
                        height="1em"
                        role="img"
                        viewBox="0 0 24 24"
                        width="1em"
                        xmlns="http://www.w3.org/2000/svg"
                      >
                        <path
                          d="M12 5.83 15.17 9l1.41-1.41L12 3 7.41 7.59 8.83 9 12 5.83zm0 12.34L8.83 15l-1.41 1.41L12 21l4.59-4.59L15.17 15 12 18.17z"
                        />
                      </svg>
                    </button>
                  </th>
                  <th
                    aria-sort="none"
                    class="table-header"
                    colspan="1"
                    role="columnheader"
                    scope="col"
                    style="padding-bottom: .5rem;"
                  >
                    <button
                      class="usa-button usa-button--unstyled position-relative"
                      style="cursor: pointer;"
                      title="Toggle SortBy"
                      type="button"
                    >
                      Role
                      <svg
                        class="usa-icon margin-left-05 position-absolute"
                        data-testid="caret--sort"
                        focusable="false"
                        height="1em"
                        role="img"
                        viewBox="0 0 24 24"
                        width="1em"
                        xmlns="http://www.w3.org/2000/svg"
                      >
                        <path
                          d="M12 5.83 15.17 9l1.41-1.41L12 3 7.41 7.59 8.83 9 12 5.83zm0 12.34L8.83 15l-1.41 1.41L12 21l4.59-4.59L15.17 15 12 18.17z"
                        />
                      </svg>
                    </button>
                  </th>
                  <th
                    aria-sort="none"
                    class="table-header"
                    colspan="1"
                    role="columnheader"
                    scope="col"
                    style="padding-bottom: .5rem;"
                  >
                    <button
                      class="usa-button usa-button--unstyled position-relative"
                      style="cursor: pointer;"
                      title="Toggle SortBy"
                      type="button"
                    >
                      Date added
                      <svg
                        class="usa-icon margin-left-05 position-absolute"
                        data-testid="caret--sort"
                        focusable="false"
                        height="1em"
                        role="img"
                        viewBox="0 0 24 24"
                        width="1em"
                        xmlns="http://www.w3.org/2000/svg"
                      >
                        <path
                          d="M12 5.83 15.17 9l1.41-1.41L12 3 7.41 7.59 8.83 9 12 5.83zm0 12.34L8.83 15l-1.41 1.41L12 21l4.59-4.59L15.17 15 12 18.17z"
                        />
                      </svg>
                    </button>
                  </th>
                  <th
                    aria-sort="none"
                    class="table-header"
                    colspan="1"
                    role="columnheader"
                    scope="col"
                    style="padding-bottom: .5rem;"
                  >
                    <button
                      class="usa-button usa-button--unstyled position-relative"
                      type="button"
                    >
                      Actions
                      <svg
                        class="usa-icon margin-left-05 position-absolute"
                        data-testid="caret--sort"
                        focusable="false"
                        height="1em"
                        role="img"
                        viewBox="0 0 24 24"
                        width="1em"
                        xmlns="http://www.w3.org/2000/svg"
                      >
                        <path
                          d="M12 5.83 15.17 9l1.41-1.41L12 3 7.41 7.59 8.83 9 12 5.83zm0 12.34L8.83 15l-1.41 1.41L12 21l4.59-4.59L15.17 15 12 18.17z"
                        />
                      </svg>
                    </button>
                  </th>
                </tr>
              </thead>
              <tbody
                role="rowgroup"
              >
                <tr
                  role="row"
                >
                  <td
                    role="cell"
                    style="white-space: normal;"
                  >
                    <div
                      class="display-flex  flex-align-center"
                    >
                      <div
                        class="display-flex flex-align-center flex-justify-center minw-4 circle-4 bg-blue-10 margin-right-1"
                        data-testid="avatar--basic"
                      >
                        JD
                      </div>
                      <div
                        class="margin-y-0"
                      >
                        <span>
                          John Doe
                        </span>
                      </div>
                    </div>
                  </td>
                  <td
                    role="cell"
                    style="white-space: normal;"
                  >
                    Model Lead
                  </td>
                  <td
                    role="cell"
                    style="white-space: normal;"
                  >
                    October 22, 2022
                  </td>
                  <td
                    role="cell"
                    style="white-space: normal;"
                  >
                    <a
                      aria-label="Edit John Doe"
                      class="usa-link margin-right-2"
                      href="/models/f11eb129-2c80-4080-9440-439cbe1a286f/collaborators/add-collaborator/61c7b30c-969d-4dd4-b13b-a5065f43be43?view=add"
                    >
                      Edit
                    </a>
                  </td>
                </tr>
              </tbody>
            </table>
            <div
              aria-live="polite"
              class="usa-sr-only usa-table__announcement-region"
            >
              Requests table reset to default sort order
            </div>
          </div>
          <div
            class="margin-top-5 display-block"
          >
            <a
              class="usa-button usa-button--outline"
              data-testid="button--back"
              href="/models/new-plan"
            >
              Back
            </a>
            <a
              class="usa-button usa-button--outline"
              data-testid="continue-to-tasklist"
              href="/models/f11eb129-2c80-4080-9440-439cbe1a286f/task-list"
            >
              Continue to Model Plan task list
            </a>
          </div>
        </div>
      </div>
    </div>
  </main>
</DocumentFragment>
`;
